dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,762 @@
+        # Cache
+        self._cache_outputs = dict()
+        # Whether we want the order of DALI execution to be reproducible
+        self._deterministic = deterministic
+        # Proxy
+        self._proxy = None
+
+    def __del__(self):
+        self.stop_thread()
+
+    @property
+    def proxy(self):
+        if not self._proxy:
+
+            class _DALIProxyCallable(_DALIProxy):
+                def __init__(self, signature, dali_input_q, deterministic):
+                    super().__init__(signature, dali_input_q, deterministic)
+
+                @_modify_signature(self._signature)
+                def __call__(self, *args, **kwargs):
+                    bound_args = self._signature.bind(self, *args, **kwargs)
+                    return self._add_sample(bound_args.arguments)
+
+            self._proxy = _DALIProxyCallable(
+                self._signature, self._dali_input_q, self._deterministic
+            )
+        return self._proxy
+
+    def _get_outputs(self, pipe_run_ref: DALIPipelineRunRef):
+        """
+        Gets the pipeline outputs for a specific batch id. It will keep reading data until the
+        right batch is found, caching the results that were not consumed until a future call
+        """
+        req_batch_id = pipe_run_ref.batch_id
+
+        # In case we haven't scheduled the iteration yet (i.e. deterministic config), do it now
+        if not pipe_run_ref.is_scheduled:
+            _nvtx.range_push(f"dali_input_q put {pipe_run_ref.batch_id}")
+            self._dali_input_q.put(pipe_run_ref)
+            pipe_run_ref.is_scheduled = True
+            _nvtx.range_pop()
+
+        # Wait for the requested output to be ready
+        req_outputs = None
+        # If the data was already read, just return it (and clear the cache entry)
+        if req_batch_id in self._cache_outputs:
+            req_outputs = self._cache_outputs[req_batch_id]
+            del self._cache_outputs[req_batch_id]
+
+        else:
+            curr_batch_id = None
+            # If not the data we are looking for, store it and keep processing until we find it
+            while req_batch_id != curr_batch_id:
+                _nvtx.range_push("dali_output_q.get")
+                curr_batch_id, curr_processed_outputs, err = self._dali_output_q.get()
+                _nvtx.range_pop()
+
+                if err is not None:
+                    raise err
+
+                if curr_batch_id == req_batch_id:
+                    req_outputs = curr_processed_outputs
+                else:
+                    self._cache_outputs[curr_batch_id] = curr_processed_outputs
+        return req_outputs
+
+    def _produce_data_impl(self, obj, cache, visited):
+        """
+        Recursive implementation of produce_data
+        """
+
+        # Check if the object has already been visited (to prevent infinite recursion)
+        obj_id = id(obj)
+        if obj_id in visited:
+            return obj
```

Comment:
I don't know it it's enough or if it can be done at all in a single pass.
It's ok if the object remains unchanged, but if you actually do change it, you should (?) return the same _transformed_ value every time you encounter this object. If the transformed object is immutable, then it's impossible to cache a reference beforehand. I think the way to go would be:
1. Create a cache that maps id(obj) to the transformed object
2. Populate it when visiting an object for the 1st time
3. After the first pass is done, recurse again and replace objects with the values from the cache.

Maybe I'm overthinking it...